plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -81,28 +80,32 @@ String toRemoteDatasetName(String projectId, String datasetName)
         requireNonNull(projectId, "projectId is null");
         requireNonNull(datasetName, "datasetName is null");
         verify(datasetName.codePoints().noneMatch(Character::isUpperCase), "Expected schema name from internal metadata to be lowercase: %s", datasetName);
-        if (caseInsensitiveNameMatching) {
-            try {
-                @Nullable String remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
-                if (remoteDatasetName == null) {
-                    // this might have been just created, reload cache
-                    // will throw on collisions to avoid ambiguity
-                    Map<String, String> mapping = stream(listDatasets(projectId))
-                            // will throw on collisions to avoid ambiguity
-                            .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
-                    remoteDatasetNames.putAll(mapping);
-                    remoteDatasetName = mapping.get(datasetName);
-                }
-                if (remoteDatasetName != null) {
-                    return remoteDatasetName;
-                }
+        if (!caseInsensitiveNameMatching) {
+            return datasetName;
+        }
+
+        try {
+            Optional<Optional<String>> optionalRemoteDatasetName = Optional.ofNullable(remoteDatasetNames.getIfPresent(datasetName));
+            if (optionalRemoteDatasetName.isPresent()) {
+                return optionalRemoteDatasetName.get().orElse(datasetName);
             }
-            catch (RuntimeException e) {
-                throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + firstNonNull(e.getMessage(), e), e);
+
+            // cache miss, reload the cache
```

Comment:
A part of me says that it would be nice to have some tests of caching. Do we have some API call statistics, to make that feasible. 
In any case I think it can be a followup.

Reply:
I agree, but there's nothing of that sort right now. I'll create an umbrella issue for BigQuery improvements and collect this there.